pageserver/src/tenant_threads.rs

```diff
@@ -1,77 +1,173 @@
 //! This module contains functions to serve per-tenant background processes,
 //! such as compaction and GC
+
 use crate::repository::Repository;
-use crate::tenant_mgr;
 use crate::tenant_mgr::TenantState;
+use crate::thread_mgr::ThreadKind;
+use crate::{tenant_mgr, thread_mgr};
 use anyhow::Result;
-use std::time::Duration;
+use once_cell::sync::OnceCell;
+use tokio::sync::mpsc::{self, Sender};
 use tracing::*;
 use utils::zid::ZTenantId;
 
 ///
-/// Compaction thread's main loop
+/// Compaction task's main loop
 ///
-pub fn compact_loop(tenantid: ZTenantId) -> Result<()> {
-    if let Err(err) = compact_loop_ext(tenantid) {
+async fn compaction_loop(tenantid: ZTenantId) -> Result<()> {
+    if let Err(err) = compaction_loop_ext(tenantid).await {
         error!("compact loop terminated with error: {:?}", err);
         Err(err)
     } else {
         Ok(())
     }
 }
 
-fn compact_loop_ext(tenantid: ZTenantId) -> Result<()> {
+async fn compaction_loop_ext(tenantid: ZTenantId) -> Result<()> {
     loop {
         if tenant_mgr::get_tenant_state(tenantid) != Some(TenantState::Active) {
             break;
         }
         let repo = tenant_mgr::get_repository_for_tenant(tenantid)?;
         let compaction_period = repo.get_compaction_period();
 
-        std::thread::sleep(compaction_period);
-        trace!("compaction thread for tenant {} waking up", tenantid);
+        tokio::time::sleep(compaction_period).await;
+        trace!("compaction loop for tenant {} waking up", tenantid);
 
         // Compact timelines
         let repo = tenant_mgr::get_repository_for_tenant(tenantid)?;
-        repo.compaction_iteration()?;
+        let compaction_result =
+            tokio::task::spawn_blocking(move || repo.compaction_iteration()).await;
```

Comment:
> tokio::task::spawn_blocking

Since you've placed this one, it would be good to notice that all function calls above this line are also blocking:
* get_tenant_state
* get_repository_for_tenant
* get_compaction_period
* get_repository_for_tenant
* and even `get_tenant_state` below, that's called for trace logging only..

This spawn looks a bit confusing tbh, I'd rather slap a TODO/warning about the blocking calls and move on without extra related changes inside, wrapping it into `spawn_blocking` somewhere above instead.